src/decoder/flac.rs

```diff
@@ -1,62 +1,302 @@
+    /// Always `Some` during normal operation and iteration.
+    reader: Option<FlacReader<R>>,
+
+    /// Buffer containing decoded samples from current block.
+    ///
+    /// Stores raw i32 samples from the current FLAC block in the decoder's native
+    /// interleaved format. Capacity is pre-allocated based on stream's maximum block size.
     current_block: Vec<i32>,
+
+    /// Number of samples per channel in current block.
+    ///
+    /// Used for calculating the correct memory layout when accessing interleaved samples.
+    /// FLAC blocks can have variable sizes, so this changes per block.
     current_block_channel_len: usize,
+
+    /// Current position within the current block.
+    ///
+    /// Tracks the next sample index to return from `current_block`. When this reaches
+    /// `current_block.len()`, a new block must be decoded.
     current_block_off: usize,
+
+    /// Number of bits per sample (8, 12, 16, 20, 24, or 32).
+    ///
+    /// Preserved from the original FLAC stream metadata and used for proper
+    /// sample conversion during iteration. FLAC supports various bit depths
+    /// including non-standard ones like 12 and 20-bit.
     bits_per_sample: u32,
+
+    /// Sample rate in Hz.
+    ///
+    /// Cached from FLAC stream metadata. FLAC supports sample rates from 1 Hz
+    /// to 655,350 Hz, though typical rates are 44.1kHz, 48kHz, 96kHz, etc.
     sample_rate: SampleRate,
+
+    /// Number of audio channels.
+    ///
+    /// FLAC supports 1 to 8 channels. Common configurations include mono (1),
+    /// stereo (2), 5.1 surround (6), and 7.1 surround (8).
     channels: ChannelCount,
+
+    /// Total duration if known from stream metadata.
+    ///
+    /// Calculated from the total sample count in FLAC metadata. `None` indicates
+    /// missing or invalid metadata, though this is rare for valid FLAC files.
     total_duration: Option<Duration>,
+
+    /// Total number of audio frames in the stream.
+    ///
+    /// Represents the total number of inter-channel samples (frames) as stored
+    /// in FLAC metadata. Used for accurate seeking and duration calculation.
+    total_samples: Option<u64>,
+
+    /// Number of samples read so far (for seeking calculations).
+    ///
+    /// Tracks the current playback position in total samples (across all channels).
+    /// Used to determine if seeking requires stream reset or can be done by
+    /// skipping forward.
+    samples_read: u64,
+
+    /// Whether the stream supports random access seeking.
+    ///
+    /// When `true`, enables backward seeking by allowing stream reset operations.
+    /// When `false`, only forward seeking (sample skipping) is allowed.
+    is_seekable: bool,
 }
 
 impl<R> FlacDecoder<R>
 where
     R: Read + Seek,
 {
-    /// Attempts to decode the data as FLAC.
-    pub fn new(mut data: R) -> Result<FlacDecoder<R>, R> {
-        if !is_flac(data.by_ref()) {
+    /// Attempts to decode the data as FLAC with default settings.
```

Comment:
note to self: got till here (dont trust git to keep my review comments unless I send them now)